### PR TITLE
Fix pool elements leftover info bug

### PR DESCRIPTION
**Pull request by guscarreon**
When these test got run:
```
	testCases := []aTest{
		{
			description:      "Blank value in put element",
			reqBody:          "{\"puts\":[{\"type\":\"xml\",\"value\":\"\"}]}",
			expectedResponse: "{\"responses\":[\"uuid\":\"\"]}",
			emptyResponses:   false,
		},
		{
			description:      "All empty body",
			reqBody:          "{}",
			expectedResponse: "{\"responses\":[]}",
			emptyResponses:   true,
		},
		{
			description:      "Empty puts arrray",
			reqBody:          "{\"puts\":[]}",
			expectedResponse: "{\"responses\":[]}",
			emptyResponses:   true,
		},
	}
```

The first one passes the test with the expected output, but the second one, wasn't getting explicitly reset after each loop just like `resps.Responses` was in line 60:

```
20 func NewPutHandler(backend backends.Backend, maxNumValues int, allowKeys bool) func(http.ResponseWriter, *http.Request, httprouter.Params) {
21 +-- 13 lines: TODO(future PR): Break this giant function apart------------------------------------------------
34     return func(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
35 +--  8 lines: body, err := ioutil.ReadAll(r.Body)-------------------------------------------------------------
43         put := putAnyRequestPool.Get().(*PutRequest)
           // not explicitly reset with make()
44         defer putAnyRequestPool.Put(put)
45
46 +-- 12 lines: err = json.Unmarshal(body, put)-----------------------------------------------------------------
58
59         resps := putResponsePool.Get().(*PutResponse)
60         resps.Responses = make([]PutResponseObject, len(put.Puts))
61         defer putResponsePool.Put(resps)
62
endpoints/put.go
```

Therefore, when a new object gets unmarshaled into an element of the `putAnyRequestPool` pool, this new element may still have some fields holding values from its last use.

```
(dlv) continue
> github.com/prebid/prebid-cache/endpoints.NewPutHandler.func3() ./put.go:47 (hits goroutine(4):1 total:1) (PC: 0x17476cb)
    42:
    43:                 put := putAnyRequestPool.Get().(*PutRequest)
    44:                 defer putAnyRequestPool.Put(put)
    45:
    46:                 err = json.Unmarshal(body, put)
=>  47:                 if err != nil {
    48:                         //http.Error(w, "Request body "+string(body)+" is not valid JSON.", http.StatusBadRequest)
    49:                         logrus.Warnf("Request body \"%s\" is not valid JSON.", string(body))
    50:                         return
    51:                 }
    52:
(dlv) p string(body)
"{\"puts\":[{\"type\":\"xml\",\"value\":\"\"}]}"
(dlv) p put.Puts[0]
github.com/prebid/prebid-cache/endpoints.PutObject {
        Type: "xml",
        TTLSeconds: 0,
        Value: encoding/json.RawMessage len: 2, cap: 8, [34,34],
        Key: "",}
(dlv) p string(put.Puts[0].Value)
"\"\""
(dlv) continue
> github.com/prebid/prebid-cache/endpoints.NewPutHandler.func3() ./put.go:47 (hits goroutine(4):2 total:2) (PC: 0x17476cb)
    42:
    43:                 put := putAnyRequestPool.Get().(*PutRequest)
    44:                 defer putAnyRequestPool.Put(put)
    45:
    46:                 err = json.Unmarshal(body, put)
=>  47:                 if err != nil {
    48:                         //http.Error(w, "Request body "+string(body)+" is not valid JSON.", http.StatusBadRequest)
    49:                         logrus.Warnf("Request body \"%s\" is not valid JSON.", string(body))
    50:                         return
    51:                 }
    52:
(dlv) p string(body)
"{}"
(dlv) p put.Puts[0]
github.com/prebid/prebid-cache/endpoints.PutObject {
        Type: "xml",                                             <=== VALUE FROM PREVIOUS TEST
        TTLSeconds: 0,                                           <=== VALUE FROM PREVIOUS TEST
        Value: encoding/json.RawMessage len: 2, cap: 8, [34,34], <=== VALUE FROM PREVIOUS TEST
        Key: "",}
(dlv) p string(put.Puts[0].Value)
"\"\""                                                           <=== VALUE FROM PREVIOUS TEST
(dlv) p put.Puts
[]github.com/prebid/prebid-cache/endpoints.PutObject len: 1, cap: 4, [
        {
                Type: "xml",
                TTLSeconds: 0,
                Value: encoding/json.RawMessage len: 2, cap: 8, [34,34],
                Key: "",},
]
```
After proposed change got implemented, the `put.Puts` value in the second test case after unmarshaling made sense.
```
> github.com/prebid/prebid-cache/endpoints.NewPutHandler.func3() ./put.go:48 (hits goroutine(21):2 total:2) (PC: 0x174771f)
    43:                 put := putAnyRequestPool.Get().(*PutRequest)
    44:                 put.Puts = make([]PutObject, 0)
    45:                 defer putAnyRequestPool.Put(put)
    46:
    47:                 err = json.Unmarshal(body, put)
=>  48:                 if err != nil {
    49:                         //http.Error(w, "Request body "+string(body)+" is not valid JSON.", http.StatusBadRequest)
    50:                         logrus.Warnf("Request body \"%s\" is not valid JSON.", string(body))
    51:                         return
    52:                 }
    53:
(dlv) p string(body)
"{}"
(dlv) p put.Puts
[]github.com/prebid/prebid-cache/endpoints.PutObject len: 0, cap: 0, []
```
Memory-wise Erik Dubbelboer's change was still an improvement. This is the benchmark of this PR's change vs the pre-Erik Dubbelboer's PR (#55 Fix proper sync.Pool usage) version that was not using pointers to pool elements:
**Before**
```
╰─➤  go test -run=XXX -bench=. -benchmem                                                                                                                                                                                                2 ↵
go: finding module for package github.com/stretchr/testify/assert
go: found github.com/stretchr/testify/assert in github.com/stretchr/testify v1.5.1
goos: darwin
goarch: amd64
pkg: github.com/prebid/prebid-cache/endpoints
BenchmarkPutHandlerLen1-12        605892              1884 ns/op             916 B/op          8 allocs/op
BenchmarkPutHandlerLen2-12        612054              2179 ns/op             955 B/op          8 allocs/op
BenchmarkPutHandlerLen4-12        572208              2084 ns/op             921 B/op          8 allocs/op
BenchmarkPutHandlerLen8-12        581739              2022 ns/op             919 B/op          8 allocs/op
PASS
ok      github.com/prebid/prebid-cache/endpoints        100.444s
```
**After both PR #55 and this PR**
```
╰─➤  go test -run=XXX -bench=. -benchmem
goos: darwin
goarch: amd64
pkg: github.com/prebid/prebid-cache/endpoints
BenchmarkPutHandlerLen1-12        633429              1899 ns/op             880 B/op          7 allocs/op
BenchmarkPutHandlerLen2-12        622588              2017 ns/op             921 B/op          7 allocs/op
BenchmarkPutHandlerLen4-12        585067              2046 ns/op             887 B/op          7 allocs/op
BenchmarkPutHandlerLen8-12        581062              1960 ns/op             888 B/op          7 allocs/op
PASS
ok      github.com/prebid/prebid-cache/endpoints        86.663s
```
Let me know your thoughts

### CODE REVIEW

**Comment by SyntaxNode** [comment]
The change looks pretty conclusive.. it's just that according to the Go doc's, it shouldn't be needed:

```
To unmarshal a JSON array into a slice, Unmarshal resets the slice length to zero and then appends each element to the slice. As a special case, to unmarshal an empty JSON array into a slice, Unmarshal replaces the slice with a new empty slice.
```

I'm a bit scared we're not seeing that behavior.

**Comment by hhhjort** [comment]
Oh, I see why we would need this change even if the documented behavior is followed exactly. We start with a request that has a non-empty `puts`. The pool object gets recycled at the end, and then gets reused to service the request `{}`. When it tries unmarshalling that, it doesn't find the `puts` object, so it won't try overwriting the `Puts` field with an empty slice. The result is an unmarshalled put that has items in the `Puts` slice. A request that looks like `{"puts":[]}` should always be empty given the documented behavior. I would be curious if the old code would fail the test if that was the only "empty" request.

**Comment by guscarreon** [comment]
>  I would be curious if the old code would fail the test if that was the only "empty" request.
The "old code" does not fail the tests that this PR adds.

**Comment by SyntaxNode** [comment]
@hhhjort Ah, ok. I'm not sure if that's the same problem we saw in production. I have a hard time believing we were sending a large amount of empty requests. That being said, there might've been some strange race condition with the slice which is resolved by this.

This looks good to me. I think we might still want to test for race conditions or continue investigating to prove it's no longer a concern.

File: endpoints/integration_test.go [hunk 1]
```diff
@@ -196,3 +198,180 @@ func TestReadinessCheck(t *testing.T) {
...
+
+		// Assertions
+		assert.Equalf(t, http.StatusOK, getResult.Code, "Description: %s \n Multi-element put failed to store:%s \n", testCases[i].description, testCases[i].elemToPut)
+		assert.Equalf(t, testCases[i].expectedStoredValue, getResult.Body.String(), "GET response error. Expected %v. Actual %v", testCases[i].expectedStoredValue, getResult.Body.String())
+	}
+}
+
+func TestEmptyPutRequests(t *testing.T) {
+	// Test case: request with more than one element in the "puts" array
+	type aTest struct {
```
Comment (SyntaxNode) [inline endpoints/integration_test.go:262]: Please add a comment indicating that order is extremely important for this test. That's not usually the assumption for table driven tests.

**Pull request merged**